File: 프로그래머스/0/181916. 주사위 게임 3/주사위 게임 3.js
function solution(a, b, c, d) {
    const l = [a, b, c, d];
    const counts = l.map(x => l.filter(y => y === x).length);
    const maxCount = Math.max(...counts);
    
    if (maxCount === 4) {
        return 1111 * a;

    } else if (maxCount === 3) {
        const p = l[counts.indexOf(3)];
        const q = l[counts.indexOf(1)];
        
        return (10 * p + q) ** 2;
    } else if (maxCount === 2) {
        const minCount = Math.min(...counts);

        if (minCount === 2) {
            const maxL = Math.max(...l);
            const minL = Math.min(...l);
            
            return (maxL + minL) * Math.abs(maxL - minL);

        } else {
            let product = 1;
            for (let i = 0; i < counts.length; i++) {
                if (counts[i] === 1) {
                    product *= l[i];
                }
            }
            return product;
        }

    } else {
        return Math.min(...l);
    }
}

File: 프로그래머스/0/181916. 주사위 게임 3/주사위 게임 3.py
from collections import Counter

def solution(a, b, c, d):
    counts = Counter([a, b, c, d])
    num_of_kinds = len(counts)
    keys = list(counts.keys())

    if num_of_kinds == 1:
        p = keys[0]
        return 1111 * p
    elif num_of_kinds == 2:
        max_freq = max(counts.values()) 

        if max_freq == 3:
            p, q = counts.most_common()[0][0], counts.most_common()[1][0]
            return (10 * p + q) ** 2
        else:
            p, q = keys[0], keys[1]
            return (p + q) * abs(p - q)

    elif num_of_kinds == 3:
        q_r_list = [k for k, v in counts.items() if v == 1]
        q, r = q_r_list[0], q_r_list[1]
        return q * r
    else:
        return min(a, b, c, d)